tests/Unit/HtmlProcessor/AbstractHtmlProcessorTest.php

```diff
@@ -629,9 +629,14 @@ public function getDomDocumentWithNormalizedHtmlRepresentsTheGivenHtml()
             "</head>\n<body>\n<br>\n</body>\n</html>\n";
         $subject = TestingHtmlProcessor::fromHtml($html);
 
-        $domDocument = $subject->getDomDocument();
+        self::assertSame("<!DOCTYPE html>
+<html><head><meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-8\"></head>
+<body>
+<br>
+</body>
 
-        self::assertSame($html, $domDocument->saveHTML());
+</html>
+", $subject->getHtml5()->saveHTML($subject->getDomDocument()));
```

Comment:
Struggled to debug this, I think there's a rogue `\r` which `masterminds/html5` is converting to `\n` so you get `\n\n` instead of `\n` in between the `</body></html>`

Reply:
This may be partly due to `masterminds/html5` [incorrectly using `PHP_EOL`](https://github.com/Masterminds/html5-php/blob/2.7.0/src/HTML5/Serializer/OutputRules.php#L456).  The line ending used should not depend on whether it happens to be running on Windows or not.

Fortunately this is only used in two places: after the `DOCTYPE`, and at the end.  Until corrected in `masterminds/html5`, we would need to replace `"\r\n"` with `"\n"` in those two places - although it probably doesn't matter, it will make testing easier.

There are some newline differences also, and a formatting-agnostic comparison will help here as elsewhere.  I found that `masterminds/html5` would always produce a `DOMDocument` with one more newline between `</body>` and `</html>` that was originally input (so an input which produces the same output is impossible).

I'm not sure we need to provide a `getHtml5` method (can the `HTML5` object remain internal?) or change this test other than dealing with the newlines issue.

Reply:
This is inconsistent behaviour on PHPs side. The bumped the minimum libxml version in PHP 7.4 but I can't see anything for PHP 7.3, but these kind of changes will be coming from the version of libxml bundled in PHP.

The issue is essentially that `masterminds/html5-php` copies over the `\n` character but `libxml` then adds it's own which results in unexpected duplication.

https://3v4l.org/vC5RX

Not much we can do about it other than change the test case.

Reply:
I had found that passing HTML through `Masterminds/html5-php` resulted in an additional newline between `</body>` and `</html>` however many were there originally, which may be related.

If you rebase with the latest `AbstractHtmlProcessorTest.php` this issue should be resolved through the tests being more forgiving of whitespace differences where they don't matter.

However, I note that I have not dealt with the issue of `Masterminds/html5-php` using `PHP_EOL` (it shouldn't be - the OS it's being run on should have no bearing on the result), so the test will still currently fail with a newline difference if you are running on Windows or Mac.